model/invoice.rb

```diff
@@ -15,58 +15,67 @@ def name
 
   def charge
+    reload # Reload to get the latest status to avoid double charging
     unless (Stripe.api_key = Config.stripe_secret_key)
-      puts "Billing is not enabled. Set STRIPE_SECRET_KEY to enable billing."
-      return
+      Clog.emit("Billing is not enabled. Set STRIPE_SECRET_KEY to enable billing.")
+      return true
     end
 
     if status != "unpaid"
-      puts "Invoice[#{ubid}] already charged: #{status}"
-      return
+      Clog.emit("Invoice already charged.") { {invoice_already_charged: {ubid: ubid, status: status}} }
+      return true
     end
 
-    if content["cost"] < Config.minimum_invoice_charge_threshold
+    amount = content["cost"].to_f.round(2)
+    if amount < Config.minimum_invoice_charge_threshold
       update(status: "below_minimum_threshold")
-      puts "Invoice[#{ubid}] cost is less than minimum charge cost: $#{content["cost"]}"
-      return
+      Clog.emit("Invoice cost is less than minimum charge cost.") { {invoice_below_threshold: {ubid: ubid, cost: amount}} }
+      return true
     end
 
-    unless (billing_info = BillingInfo[content.dig("billing_info", "id")])
-      puts "Invoice[#{ubid}] doesn't have billing info"
-      return
+    if (billing_info = BillingInfo[content.dig("billing_info", "id")]).nil? || billing_info.payment_methods.empty?
+      Clog.emit("Invoice doesn't have billing info.") { {invoice_no_billing: {ubid: ubid}} }
+      return false
     end
 
-    payment_methods = billing_info.payment_methods_dataset.order(:order).all
-
-    payment_methods.each do |payment_method|
-      amount = content["cost"].to_f.round(2)
-      payment_intent = Stripe::PaymentIntent.create({
-        amount: (amount * 100).to_i, # 100 cents to charge $1.00
-        currency: "usd",
-        confirm: true,
-        off_session: true,
-        customer: billing_info.stripe_id,
-        payment_method: payment_method.stripe_id
-      })
-
-      if payment_intent.status == "succeeded"
-        puts "Invoice[#{ubid}] charged with PaymentMethod[#{payment_method.ubid}] for $#{amount}"
-        self.status = "paid"
-        content.merge!({
-          "payment_method" => {
-            "id" => payment_method.id,
-            "stripe_id" => payment_method.stripe_id
-          },
-          "payment_intent" => payment_intent.id
+    errors = []
+    billing_info.payment_methods_dataset.order(:order).each do |pm|
+      begin
+        payment_intent = Stripe::PaymentIntent.create({
+          amount: (amount * 100).to_i, # 100 cents to charge $1.00
+          currency: "usd",
+          confirm: true,
+          off_session: true,
+          customer: billing_info.stripe_id,
+          payment_method: pm.stripe_id
         })
-        save(columns: [:status, :content])
-        return payment_intent.id
+      rescue Stripe::CardError => e
+        Clog.emit("Invoice couldn't charged.") { {invoice_not_charged: {ubid: ubid, payment_method: pm.ubid, error: e.message}} }
+        errors << e.message
```

Comment:
Is this system error or user error. If it is system, then maybe we shouldn't put it to the errors array that we eventually send to user.